In [ ]:
import numpy as np
import pandas as pd

**Data Reading**

In [ ]:
df_cyber = pd.read_csv('/content/hateXplain.csv')

In [ ]:
df_cyber

post_id  annotator_id       label  \
0      1179055004553900032_twitter             1      normal   
1      1179055004553900032_twitter             2      normal   
2      1179055004553900032_twitter             3      normal   
3      1179063826874032128_twitter             1      normal   
4      1179063826874032128_twitter             2      normal   
...                            ...           ...         ...   
60439                  9992513_gab           253  hatespeech   
60440                  9992513_gab           222   offensive   
60441                  9998729_gab           200  hatespeech   
60442                  9998729_gab           202   offensive   
60443                  9998729_gab           207   offensive   

                       target  \
0                         NaN   
1                         NaN   
2                         NaN   
3                         NaN   
4                         NaN   
...                       ...   
60439                   Asian   
60440                   Asian   
60441          African, Islam   
60442           Islam, Jewish   
60443  African, Islam, Jewish   

                                             post_tokens  
0      i dont think im getting my baby them white 9 h...  
1      i dont think im getting my baby them white 9 h...  
2      i dont think im getting my baby them white 9 h...  
3      we cannot continue calling ourselves feminists...  
4      we cannot continue calling ourselves feminists...  
...                                                  ...  
60439  it is awful look at world demographics asians ...  
60440  it is awful look at world demographics asians ...  
60441  the jewish globalist elite have only imported ...  
60442  the jewish globalist elite have only imported ...  
60443  the jewish globalist elite have only imported ...  

[60444 rows x 5 columns]

In [ ]:
df_cyber.shape

(60444, 5)

In [ ]:
df_cyber.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60444 entries, 0 to 60443
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   post_id       60444 non-null  object
 1   annotator_id  60444 non-null  int64 
 2   label         60444 non-null  object
 3   target        39252 non-null  object
 4   post_tokens   60444 non-null  object
dtypes: int64(1), object(4)
memory usage: 2.3+ MB


In [ ]:
df_cyber.describe()

annotator_id
count  60444.000000
mean     121.987145
std       95.952225
min        1.000000
25%       13.000000
50%      199.000000
75%      211.000000
max      253.000000

In [ ]:
df_cyber = df_cyber.dropna(subset = ['post_id', 'label', 'post_tokens'])

In [ ]:
group_df = df_cyber.groupby('post_id')
label = group_df['label'].agg(lambda x:x.value_counts().index[0]).to_frame().reset_index()

if the post_id is the identifier of social media posts and multiple comments belong to the same post this will group all comments by their respective to post_id

This code groups the dataset df_cyber by post_id and finds the most frequent label (likely representing a type of cyberbullying) for each group of comments.

In [ ]:
df_cyber = pd.merge(left = df_cyber[['post_id', 'post_tokens']],
                    right = label, on = 'post_id',
                    how = 'inner').drop_duplicates(subset = ['post_id']).reset_index(drop=True)

In [ ]:
df_cyber

post_id  \
0      1179055004553900032_twitter   
1      1179063826874032128_twitter   
2      1178793830532956161_twitter   
3      1179088797964763136_twitter   
4      1179085312976445440_twitter   
...                            ...   
20143                  9989999_gab   
20144                  9990225_gab   
20145                  9991681_gab   
20146                  9992513_gab   
20147                  9998729_gab   

                                             post_tokens       label  
0      i dont think im getting my baby them white 9 h...      normal  
1      we cannot continue calling ourselves feminists...      normal  
2                          nawt yall niggers ignoring me      normal  
3      <user> i am bit confused coz chinese ppl can n...  hatespeech  
4      this bitch in whataburger eating a burger with...  hatespeech  
...                                                  ...         ...  
20143  if ur still on twitter tell carlton i said his...   offensive  
20144  when i first got on here and said i hate trump...   offensive  
20145  was macht der moslem wenn der zion gegen seine...      normal  
20146  it is awful look at world demographics asians ...  hatespeech  
20147  the jewish globalist elite have only imported ...   offensive  

[20148 rows x 3 columns]

**Pre-Processing text**

In [ ]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

import nltk
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
def preprocess_text(text):
    # Lowercasing
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)

    #tokenize
    tokens = word_tokenize(text)

    # stopwords
    stop_words = set(stopwords.words('english'))
    filtered_tokens = [word for word in tokens if word not in stop_words]


    preprocessed_text = ' '.join(filtered_tokens)
    return preprocessed_text

def remove_numbers(text):
    text=''.join([i for i in text if not i.isdigit()])
    return text

In [ ]:
df_cyber['post_tokens'] = df_cyber['post_tokens'].apply(preprocess_text)

In [ ]:
df_cyber['post_tokens'] = df_cyber['post_tokens'].apply(preprocess_text)

In [ ]:
df_cyber

post_id  \
0      1179055004553900032_twitter   
1      1179063826874032128_twitter   
2      1178793830532956161_twitter   
3      1179088797964763136_twitter   
4      1179085312976445440_twitter   
...                            ...   
20143                  9989999_gab   
20144                  9990225_gab   
20145                  9991681_gab   
20146                  9992513_gab   
20147                  9998729_gab   

                                             post_tokens       label  
0      dont think im getting baby white 9 two white j...      normal  
1      continue calling feminists rights womxn arent ...      normal  
2                             nawt yall niggers ignoring      normal  
3      user bit confused coz chinese ppl access twitt...  hatespeech  
4      bitch whataburger eating burger top bun holdin...  hatespeech  
...                                                  ...         ...  
20143  ur still twitter tell carlton said alcoholic d...   offensive  
20144  first got said hate trump considered troll peo...   offensive  
20145  macht der moslem wenn der zion gegen seinen pr...      normal  
20146  awful look world demographics asians fucking e...  hatespeech  
20147  jewish globalist elite imported million muslim...   offensive  

[20148 rows x 3 columns]

**Pre-Processing Label**

In [ ]:
from sklearn.preprocessing import LabelEncoder

Label_Encoder = LabelEncoder()
df_cyber['label_encoder'] = Label_Encoder.fit_transform(df_cyber['label'])

In [ ]:
df_cyber

post_id  \
0      1179055004553900032_twitter   
1      1179063826874032128_twitter   
2      1178793830532956161_twitter   
3      1179088797964763136_twitter   
4      1179085312976445440_twitter   
...                            ...   
20143                  9989999_gab   
20144                  9990225_gab   
20145                  9991681_gab   
20146                  9992513_gab   
20147                  9998729_gab   

                                             post_tokens       label  \
0      dont think im getting baby white 9 two white j...      normal   
1      continue calling feminists rights womxn arent ...      normal   
2                             nawt yall niggers ignoring      normal   
3      user bit confused coz chinese ppl access twitt...  hatespeech   
4      bitch whataburger eating burger top bun holdin...  hatespeech   
...                                                  ...         ...   
20143  ur still twitter tell carlton said alcoholic d...   offensive   
20144  first got said hate trump considered troll peo...   offensive   
20145  macht der moslem wenn der zion gegen seinen pr...      normal   
20146  awful look world demographics asians fucking e...  hatespeech   
20147  jewish globalist elite imported million muslim...   offensive   

       label_encoder  
0                  1  
1                  1  
2                  1  
3                  0  
4                  0  
...              ...  
20143              2  
20144              2  
20145              1  
20146              0  
20147              2  

[20148 rows x 4 columns]

In [ ]:
df_cyber[['label','label_encoder']].value_counts()

label       label_encoder
normal      1                8153
hatespeech  0                6234
offensive   2                5761
Name: count, dtype: int64

**Pre-Processing text for BertTokenizer**

In [ ]:
import torch #for building and training machine learning models
from torch.utils.data import DataLoader, Dataset # Loading the data efficiently during model training
from transformers import BertTokenizer, BertForSequenceClassification,AdamW #Tokenize the input text specifically for BERT
# A pre-trained BERT model specifically designed for sequence classification tasks, such as sentiment analysis or text classification
# weight decay
from transformers import get_linear_schedule_with_warmup #create learning rate scheduler
import torch.nn.functional as F

In [ ]:
from transformers import BertTokenizer

posts = df_cyber['post_tokens']
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased') #tokenizer treats the uppercase and lowercase the letters the same
encoded_post = tokenizer(posts.tolist(),padding=True, truncation=True, return_tensors='pt')
print(encoded_post['input_ids'])

#input_ids = encoded_post['input_ids']
#for i, input_id in enumerate(input_ids):
#   print(f"Post {i+1}:")
#   print(input_id.tolist())
#   print("\n" + "-"*50 + "\n")'''

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

tensor([[  101,  2123,  2102,  ...,     0,     0,     0],
        [  101,  3613,  4214,  ...,     0,     0,     0],
        [  101,  6583, 26677,  ...,     0,     0,     0],
        ...,
        [  101, 24532,  2102,  ...,     0,     0,     0],
        [  101,  9643,  2298,  ...,     0,     0,     0],
        [  101,  3644,  3795,  ...,     0,     0,     0]])


* Post variable that extracts the post_tokens contains social media comments
* It loads a pre-trained BERT tokenizer which converts raw text into tokens and lowercasing all texts.
* tokenizes the posts into BERT-compatible token IDs, applies padding to make all sequences the same length, truncates sequences that exceed the model's maximum input size, and returns the output as PyTorch tensors
* tokenizes the posts into BERT-compatible token IDs, applies padding to make all sequences the same length, truncates sequences that exceed the model's maximum input size, and returns the output as PyTorch tensors

**The above output is all about the BERT tokenizer where each sentences are broken down into smaller pieces and then convert each piece into a number that model can understand.**


The sentences are passed into BertTokenizer and each word is converted into a token id

BERT adds special tokens like [CLS] at the beginning and [SEP] at the end of each sentence

The 0s above are the padding length to make the sequence length of the sentence uniform

In [ ]:
input_ids = encoded_post['input_ids']
attention_mask = encoded_post['attention_mask'] #tells about which tokens are actual words and which are padding (zeroes)

In [ ]:
print(input_ids)
print(attention_mask)

tensor([[  101,  2123,  2102,  ...,     0,     0,     0],
        [  101,  3613,  4214,  ...,     0,     0,     0],
        [  101,  6583, 26677,  ...,     0,     0,     0],
        ...,
        [  101, 24532,  2102,  ...,     0,     0,     0],
        [  101,  9643,  2298,  ...,     0,     0,     0],
        [  101,  3644,  3795,  ...,     0,     0,     0]])
tensor([[1, 1, 1,  ..., 0, 0, 0],
        [1, 1, 1,  ..., 0, 0, 0],
        [1, 1, 1,  ..., 0, 0, 0],
        ...,
        [1, 1, 1,  ..., 0, 0, 0],
        [1, 1, 1,  ..., 0, 0, 0],
        [1, 1, 1,  ..., 0, 0, 0]])


**Dataset and DataLoader**

In [ ]:
class CyberBullyingDataset(Dataset):
    def __init__(self, input_ids, attention_mask, labels):
        self.input_ids = input_ids
        self.attention_mask = attention_mask
        self.labels = labels #indicates whether the post is cyberbullying or not

    def __len__(self):
        return len(self.labels)

    def __getitem__(self, idx):
        return {
            'input_ids': self.input_ids[idx],
            'attention_mask': self.attention_mask[idx],
            'labels': self.labels[idx]
        }

The code defines a custom Pytorch dataset called Cyberbullying which will handle and organize input data for a cyberbullying detection model.

It stores the tokenized inputs, attention_mask and corresponding labels.

The class allows for efficient indexing and retrieval of data samples during model training.  

In [ ]:
label_tensor = torch.tensor(df_cyber['label_encoder'].values)
dataset = CyberBullyingDataset(input_ids,attention_mask, label_tensor)
dataloader = DataLoader(dataset, batch_size=32, shuffle=True)

The dataset - This object now holds the processed data which includes the tokenized post, attention_masks and labels

Dataloader - training the model efficiently

**Model Making for BERT classification**

In [ ]:
model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=3)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu') # to speed up the training process of machine learning models
model.to(device)

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e

In [ ]:
optimizer = AdamW(model.parameters(), lr=2e-5, eps=1e-8)
num_epochs = 15
# Total number of training steps
total_steps = len(dataloader) * num_epochs

scheduler = get_linear_schedule_with_warmup(optimizer,
                                            num_warmup_steps=0,
                                            num_training_steps=total_steps)

/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:591: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


Adamw - A optimizer that adjusts the model parameters based on the error it makes and also it helps the model to prevent from overfitting.

lr = 2e-5 - This sets the learning rate which controls how much the optimizer should change the model's parameters.

eps=1e-8: This is a tiny number added to prevent division by zero errors during optimization. It ensures numerical stability.

get_linear_schedule_with_warmup: This function creates a schedule where the learning rate starts at zero, gradually increases to the initial learning rate, and then linearly decreases to zero by the end of training

In [ ]:
epochs = 10  # looks at the entire dataset 10 times

for epoch in range(epochs):
    model.train()
    total_loss = 0 #(how far the model's prediction are from actual labels)
    correct_predictions = 0 # it counts how many prediction the models get correct during epoch

    for batch in dataloader:
        optimizer.zero_grad() # cleaning previous gradients to ensure that model updates correctly based on current batch

        #Moving the data to device
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)


        outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
        loss = outputs.loss # The loss is calculated based on how far off the model's predictions are from the true labels
        logits = outputs.logits# These are the raw predictions made by the model before applying any activation function


        total_loss += loss.item()

        #BACKPROPOGATION AND OPTIMIZATION
        loss.backward()
        optimizer.step()
        scheduler.step()


        preds = torch.argmax(F.softmax(logits, dim=1), dim=1)# Converts the model's raw predictions (logits) into final class predictions
        correct_predictions += torch.sum(preds == labels)# Counts how many of the model's predictions match the true labels
                                                        # and adds them to the total number of correct predictions.

    avg_loss = total_loss / len(dataloader)
    accuracy = correct_predictions.double() / len(dataloader.dataset)

    print(f'Epoch {epoch+1}, Loss: {avg_loss}, Accuracy: {accuracy}')


Epoch 1, Loss: 0.8321621400023264, Accuracy: 0.6197637482628549
Epoch 2, Loss: 0.6919472027865667, Accuracy: 0.7047845940043677
Epoch 3, Loss: 0.5752970451400393, Accuracy: 0.7644431209053008
Epoch 4, Loss: 0.43096397270758946, Accuracy: 0.8311494937462776
Epoch 5, Loss: 0.2839618309031403, Accuracy: 0.8963172523327377
Epoch 6, Loss: 0.18916506778507952, Accuracy: 0.931705380186619
Epoch 7, Loss: 0.13189395548598398, Accuracy: 0.9538912050823902
Epoch 8, Loss: 0.09304739298919837, Accuracy: 0.9692277149096684
Epoch 9, Loss: 0.07093124184979215, Accuracy: 0.97548143736351
Epoch 10, Loss: 0.05331782371411839, Accuracy: 0.9815366289458011


Loss represents how wrong the model's prediction are. If lower loss means the model is doing great and the goal is to minimize loss during training.

**Prediction using BERT**

In [ ]:
df_final_cyber = pd.read_csv('/content/final_hateXplain.csv')

In [ ]:
df_final_cyber.head()

comment       label     Race  \
0                     0 u0 lmao wow fuck you too 😂 😂      normal  No_race   
1  1 0 th floor maybe wow cnn with the fakenews t...   offensive  No_race   
2  1 0 yrs <number> white women raped by niggers ...  hatespeech  African   
3  1 2 h ago ching chong accepted your friend req...   offensive    Asian   
4  1 8 th century mayhem and lawlessness had noth...      normal  No_race   

       Religion     Gender Sexual Orientation Miscellaneous  
0  Nonreligious  No_gender     No_orientation           NaN  
1  Nonreligious  No_gender     No_orientation         Other  
2  Nonreligious  No_gender     No_orientation           NaN  
3  Nonreligious  No_gender     No_orientation           NaN  
4  Nonreligious  No_gender     No_orientation           NaN

**Preprocessing text**

In [ ]:
df_final_cyber['comment'] = df_final_cyber['comment'].apply(preprocess_text)
df_final_cyber['comment'] = df_final_cyber['comment'].apply(remove_numbers)

**Preprocessing Label**

In [ ]:
from sklearn.preprocessing import LabelEncoder

df_final_cyber['label_encoder'] = Label_Encoder.fit_transform(df_final_cyber['label'])

In [ ]:
df_final_cyber[['label','label_encoder']].value_counts()

label       label_encoder
normal      1                7818
hatespeech  0                6484
offensive   2                5807
Name: count, dtype: int64

In [ ]:
test_comments = df_final_cyber['comment'].tolist()

test_encodings = tokenizer(test_comments, padding=True, truncation=True, return_tensors='pt')

test_labels = torch.tensor(df_final_cyber['label_encoder'].values)

In [ ]:
test_dataset = CyberBullyingDataset(test_encodings['input_ids'], test_encodings['attention_mask'], test_labels)
test_dataloader = DataLoader(test_dataset, batch_size=32, shuffle=False)

In [ ]:
model.eval()
predictions = []
true_labels = []

with torch.no_grad():
    for batch in test_dataloader:
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)

        # Forward pass
        outputs = model(input_ids=input_ids, attention_mask=attention_mask)
        logits = outputs.logits


        preds = torch.argmax(logits, dim=1)
        predictions.extend(preds.cpu().numpy())
        true_labels.extend(labels.cpu().numpy())

In [ ]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(true_labels, predictions)
print(f'The accuracy score : {accuracy:.4f}')

The accuracy score : 0.9304


The model achieved 93.04% accuracy, meaning it correctly classified cyberbullying posts 93 out of 100 times

**Summary**

In the cyberbullying detection project, BERT is used to understand the meaning of social media comments and classify whether they contain cyberbullying based on categories like race, gender, or religion.

How BERT works in this project:

Text Input: BERT takes the social media comments (text) and tokenizes them (splits them into words and subwords that the model can understand).

Contextual Understanding: BERT looks at each word in the comment in relation to other words to understand the full context. For example, it can tell the difference between "run fast" and "fast food," where "fast" has different meanings.

Classification: After processing the comment, BERT outputs a prediction: whether the comment is bullying or not, and if so, which category (e.g., based on race, gender, etc.).

Real-world example:

Imagine someone posts: "You're such a loser, nobody likes you because of where you're from."

BERT looks at the entire sentence and understands that this message might involve cyberbullying related to race or nationality.
After processing the text, BERT predicts this as cyberbullying in the "race" category.
In simple terms, BERT helps detect harmful or offensive comments by analyzing the words and their context in the sentence to determine if it's bullying and the type of bullying involved.